src/ImmutableProperties.php

```diff
@@ -0,0 +1,120 @@
+use function spl_object_hash;
+use function sprintf;
+
+trait ImmutableProperties
+{
+    /** @return string[][]|ReflectionProperty[][] */
+    protected function &reflectionProperty(): array
+    {
+        static $reflectionProperty = [];
+
+        if (! array_key_exists(spl_object_hash($this), $reflectionProperty)) {
+            $reflectionProperty[spl_object_hash($this)] = [];
+        }
+
+        return $reflectionProperty[spl_object_hash($this)];
+    }
+
+    public function __construct()
+    {
+        $this->init();
+    }
+
+    public function init(): void
+    {
+        $reflection = new ReflectionObject($this);
+        $store = &$this->reflectionProperty();
+        $defaultProperties = $reflection->getDefaultProperties();
+
+        foreach ($reflection->getProperties(ReflectionProperty::IS_PUBLIC) as $reflectionProperty) {
+            $propertyName = $reflectionProperty->name;
+            /** @var \ReflectionType $type */
+            $type = $reflectionProperty->getType();
+
+            $store[$propertyName] = [
+                'reflection' => $reflectionProperty,
+                'value' => $defaultProperties[$propertyName] ?: null,
+                'isInitialized' => isset($defaultProperties[$propertyName]) ? true : $reflectionProperty->isInitialized($this),
+            ];
+
+            if (isset($defaultProperties[$propertyName]) || $type->allowsNull()) {
+                unset($this->{$propertyName});
+            }
+        }
+    }
+
+    public function __set(string $prop, $value)
+    {
+        $ref = &$this->reflectionProperty();
+        $property = $ref[$prop];
+
+        if ($property['isInitialized']) {
+            // @TODO throws specific exception?
+            throw new RuntimeException(sprintf('Cannot reassign value to property "%s"', $prop));
+        }
+
+        /** @var ReflectionProperty $reflection */
+        $reflection = $property['reflection'];
+
+        if ($reflection->hasType()) {
+            /** @var ReflectionType $type */
+            $type            = $reflection->getType();
+            $expectedType    = $type->getName();
+            $actualValueType = is_object($value) ? get_class($value) : gettype($value);
+
+            if (is_object($value) && interface_exists($expectedType) && !in_array($expectedType, class_implements($actualValueType), false)) {
+                throw TypeErrorExceptionFactory::fromWrongType($this, $prop, $expectedType, $actualValueType);
+            }
+
+            if (($expectedType !== $actualValueType && $type->isBuiltin())
+                && (null !== $value && !$type->allowsNull())) {
+                throw TypeErrorExceptionFactory::fromWrongType($this, $prop, $expectedType, $actualValueType);
+            }
+        }
+
+        $ref[$prop] = [
+            'reflection' => $property['reflection'],
+            'value' => $value,
+            'isInitialized' => true,
+        ];
+    }
+
+    public function __get(string $prop)
```

Comment:
Not really needed if we use the real properties

Reply:
Also: currently breaks if trying to work with `private` and `protected` properties